src/EditorFeatures/Text/Extensions.TextBufferContainer.cs

```diff
@@ -47,11 +47,27 @@ public static TextBufferContainer From(ITextBuffer buffer)
                     throw new ArgumentNullException(nameof(buffer));
                 }
 
-                return s_textContainerMap.GetValue(buffer, s_createContainerCallback);
+                var weakReference = s_textContainerMap.GetValue(buffer, s_createContainerCallback);
+                var container = weakReference.GetTarget();
+
+                if (container != null)
+                {
+                    return container;
+                }
+
+                // The container has been garbage collected, so we need to create a new one; in this case
+                // we need to ensure we don't potentially create two and hand out more than one container
+                // for the same buffer, so we'll take a lock on the WeakReference to serialize.
+                lock (weakReference)
+                {
+                    container = new TextBufferContainer(buffer);
```

Comment:
❗ This needs a second check for `GetTarget() != null`

Reply:
Yep, that's why I moved it back to draft.